AppController/lib/haproxy.rb

```diff
@@ -321,4 +347,121 @@ def self.initialize_config()
+      "socat stdio unix-connect:/etc/haproxy/stats | grep #{app_name}")
+    Djinn.log_debug("HAProxy raw stats: #{monitoring_info}")
+
+    if monitoring_info.empty?
+      Djinn.log_warn("Didn't see any monitoring info - #{app_name} may not " +
+        "be running.")
+      return :no_change, :no_change, :no_backend
+    end
+
+    monitoring_info.each_line { |line|
+      parsed_info = line.split(',')
+      # no request info here
+      next if parsed_info.length < TOTAL_REQUEST_RATE_INDEX
+
+      service_name = parsed_info[SERVICE_NAME_INDEX]
+
+      if service_name == "FRONTEND"
+        total_requests_seen = parsed_info[TOTAL_REQUEST_RATE_INDEX].to_i
+        time_requests_were_seen = Time.now.to_i
+        Djinn.log_debug("#{app_name} #{service_name} Requests Seen " +
+          "#{total_requests_seen}")
+      end
+
+      if service_name == "BACKEND"
+        total_req_in_queue = parsed_info[REQ_IN_QUEUE_INDEX].to_i
+        Djinn.log_debug("#{app_name} #{service_name} Queued Currently " +
+          "#{total_req_in_queue}")
+      end
+    }
+
+    return total_requests_seen, total_req_in_queue, time_requests_were_seen
+  end
+
+
+  # This method returns the list of locations associated with a specific
+  # application.
+  #
+  # Args:
+  #   app: A String containing the application ID.
+  #   down: A Boolean, when true we return only the AppServers that are
+  #     marked as DOWN, otherwise we return them all.
+  # Returns:
+  #   An Array of AppServers (ip:port).
+  def self.list_servers(app, down=false)
+    full_app_name = "gae_#{app}"
+
+    locations = []
+    servers = Djinn.log_run("echo \"show stat\" | socat stdio " +
+      "unix-connect:/etc/haproxy/stats | grep \"#{full_app_name}\"")
+    servers.each_line{ |line|
+      parsed_info = line.split(',')
+      next if parsed_info[SERVICE_NAME_INDEX] == "FRONTEND"
+      next if parsed_info[SERVICE_NAME_INDEX] == "BACKEND"
+      next if down && parsed_info[SERVER_STATUS] != "DOWN"
+      locations << parsed_info[SERVICE_NAME_INDEX].sub(/^#{full_app_name}-/,'')
+    }
+    Djinn.log_debug("Haproxy: found these appserver for #{app}: #{locations}.")
+    return locations
+  end
+
+  # This method set the weight of the specified server (which needs to be
+  # in haproxy) to 0 (that is no more new requests will be sent to the
+  # server) and return the number of current sessions handled by such
+  # server.
+  #
+  # Args:
+  #   app: A String containint the application ID.
+  #   location: A String with the host:port of the running appserver.
+  # Returns:
+  #   An Integer containing the number of current sessions, or -1 if the
+  #     server is not known by haproxy.
+  def self.ensure_no_pending_request(app, location)
+    full_app_name = "gae_#{app}"
+    appserver = "#{full_app_name}-#{location}"
+    Djinn.log_debug("We will put 0% weight to #{appserver}.")
+
+    # Let's set the weight to 0, and check if the server is actually know.
```

Comment:
thanks, that would be known. The check makes sure we have a sane value back, not 'No such server'.